agentendpoint/agentendpoint.go

```diff
@@ -373,16 +370,27 @@ func (c *Client) WaitForTaskNotification(ctx context.Context) {
 					c.Close()
 					return
 				}
-				clog.Errorf(ctx, "Error waiting for task: %v", err)
-				sleep := 5 * time.Second
+
 				if errors.Is(err, errResourceExhausted) {
-					sleep = retryutil.RetrySleep(resourceExhausted, 5)
 					resourceExhausted++
+					sleep = retryutil.RetrySleep(resourceExhausted, 5)
 				} else {
-					resourceExhausted = 1
+					// Retry any other errors with a modest backoff. Only retry up to 10
+					// times, at that point return, the client will be recreated during the next
+					// cycle.
+					errs++
+					clog.Warningf(ctx, "Error waiting for task (attempt %d of 10): %v", errs, err)
+					resourceExhausted = 0
+					if errs > 10 {
```

Comment:
Would it be easier to reason about failures if this used a time-based threshold instead, for example, giving up after 2 minutes (I think this gives up after around 1 minute but can vary a lot depending on the random jitter of the retry delay).

Reply:
We just want to retry enough for very temporary issues to resolve themselves. Timing is very complicated as we also have the request timeouts to deal with. Time based might be easier to read if all we cared about was time since first failure, but would be more complicated to implement and ensure we try at least x number of times, so I'd end up implementing both time based on retry based. Since min retry time for 10 errors is over 6min that felt like a good point to just quit and wait for the next reload (max 10mins)